adding basic auto-scaling functionality for raw and epochs classes

choldgraf opened the pull request:
This is a first-pass PR at adding auto-scaling functionality for plotting raw / epoched data. The basic change is to include a "compute_scalings" function in the viz.utils module. This takes a scalings dictionary, and for any value that is 'auto', it will use the data instance provided to auto-calculate the scaling. Alternatively you can supply only the string "auto" and it'll auto-compute for all channel types in the data. Then it goes on to plotting as normal.

There are some weird things that I did to make iterating through channel types easier (e.g., I've got a hard-coded list of channel types but maybe that exists elsewhere, I also wrote a function to return a list of the channel type for all channels, as well as a dictionary of channel type: ixs, but maybe that's already been done elsewhere?)

Suggestions etc are welcome! Here's the basic idea of what this does:

```python
scalings_def = dict(eeg='auto', grad='auto')
scalings_raw = mne.viz.utils.compute_scalings(scalings_def, raw)
scalings_ep = mne.viz.utils.compute_scalings(scalings_def, epochs)
scalings_all = mne.viz.utils.compute_scalings('auto', raw)
print('{}\n\n{}\n\n{}\n\n{}'.format(scalings_def, scalings_raw, scalings_ep, scalings_all))
```
```
{'eeg': 'auto', 'grad': 'auto'}

{'eeg': 7.1589975858543653e-05, 'grad': 1.928710961556987e-11}

{'eeg': 3.6195661505710243e-05, 'grad': 3.506747202830885e-11}

{'ecg': None, 'bio': None, 'emg': None, 'stim': 0.0, 'eog': 1.7976379496076333e-05, 'ref_meg': None, 'resp': None, 'misc': None, 'seeg': None, 'chpi': None, 'mag': 1.3265991686001318e-12, 'syst': None, 'eeg': 7.1589975858543653e-05, 'grad': 1.928710961556987e-11, 'ias': None, 'exci': None}
```
originally comes from #2253

kingjr commented on mne/viz/utils.py:
```diff
@@ -30,6 +30,10 @@
 
+_channel_types = ['eeg', 'seeg', 'eog', 'ecg', 'emg', 'ref_meg', 'stim',
+                  'resp', 'misc', 'chpi', 'syst', 'ias', 'exci', 'bio']
+_meg_types = ['mag', 'grad']
```
shouldn't this be (already?) in mne.channels?

kingjr commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1036,102 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+    scalings : dict
+        The scalings for each channel type. If any values are
+        'auto', this will automatically compute a reasonable
+        scaling for that channel type.
```
Defaults to ...

choldgraf replied:
I put in a sentence explaining that values that aren't "auto" won't be changed, think that's OK?

kingjr replied:
Sure. 

I think eventually we'll have to set a `default` docstring convention for html formatting.

kingjr commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1036,102 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+        'auto', this will automatically compute a reasonable
+        scaling for that channel type.
+    inst : instance of Raw or Epochs
+        The data you wish to compute scalings for. If Epochs,
```
The data for which you wish to compute the scalings.

choldgraf replied:
I didn't realize that MNE was anti-hanging prepositions ;)

kingjr replied:
I just really enjoy exposing my knowledge of English grammar to the world. It's in response to the frustration Reviewers #2 generate, when asking for improving the language of my manuscripts.

kingjr commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1036,102 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+    Returns
+    -------
+    scalings : dict
+        A scalings dictionary with updated scalings
```
A scaling dictionary with updated values.

choldgraf replied:
done

kingjr commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1036,102 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+
+    if scalings == 'auto':
+        # If we want to auto-compute everything
+        scalings = {i_type: 'auto' for i_type in _channel_types + _meg_types}
```
is this python3 compatible? I thought we had to use `dict()` necessarily, but I may be wrong

choldgraf replied:
hmm, I'm not sure. I just tried a dictionary comprehension in python 3 and it worked fine...I can change this if it'll be an issue

jona-sassenhagen replied:
This is standard Py3 dict comp.

I'm fairly sure it will also work for 2.6.

But I've been repeatedly told we shouldn't do it like this here, although I don't know why either.

Eric89GXL replied:
I don't think it works in 2.6. But I think you can do `dict((key, value) for key, value in ...)` for all Pythons.

kingjr commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1036,102 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+        data = np.hstack(inst._data)
+
+    # Arbitrary compression factor for scaling
+    scale_factor = 5. if isinstance(inst, _BaseRaw) else 10.
```
can you explain this?

choldgraf replied:
totally ad-hoc and messy. I noticed that scaling parameters would tend to be a little different between epochs and raw. In particular, with Epochs there usually more "stuff going on" and the plots looked busier. So it seemed like a different scaling factor for each might work better, but maybe that's too hacky...

kingjr replied:
I find it a bit too hacky yes...

SNR of epochs is higher, so the overall STD drops as compared to raw, but this should be captured by the `threshold = n * mad(data)`.

Thinking of this, we should probably use `np.percentile(data, n)` (with `n=99`) instead of the mad. It'll be cleaner to manipulate later on I think.

choldgraf replied:
that's a good idea - latest push uses 99.5th percentile (also pretty much an arbitrary number but I do it instinctively from doing permutation tests :P ) instead of standard deviations etc...that seems to work reasonably well for different kinds of channels, as well as both for Raw + Epochs.

kingjr commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1036,102 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+    inds = {}
+
+    # First MEG because there are special kws for it
+    for t in _meg_types:
```
t -> this_type

(we shouldn't use single letter, it conflicts with the debugger)

agramfort commented:
have a look at the _picks_by_type function in pick.py and maybe _PICK_TYPES_DATA_DICT, _PICK_TYPES_KEYS and _DATA_CH_TYPES_SPLIT in pick.py

choldgraf commented:
Think I addressed people's comments. Using percentile now instead of absolute median deviation, also updated dictionaries to be 2.6 compatible

choldgraf commented:
Also hanging participle has been fixed @kingjr, I'll make sure to bring my copy of Strunk and White for my next PR :)

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1032,71 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
     return fig
+
+
+def compute_scalings(scalings, inst):
```
make private

choldgraf replied:
done

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1032,71 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+    scalings : dict
+        A scalings dictionary with updated values
+    """
+    from copy import deepcopy
```
don't nest this

choldgraf replied:
done

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1032,71 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+    from copy import deepcopy
+    from mne.io.base import _BaseRaw
+    from mne.io.pick import _picks_by_type
+    from mne.epochs import _BaseEpochs
```
use relative imports

choldgraf replied:
done

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1032,71 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+        # If we want to auto-compute everything
+        scalings = dict((i_type, 'auto') for i_type in unique_ch_types)
+    if not isinstance(scalings, dict):
+        raise ValueError('scalings must be a dictionary of ch_type: val pairs')
```
, not %s' % type(scalings)

choldgraf replied:
done

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1032,71 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+    else:
+        data = inst._data
+    if isinstance(inst, _BaseEpochs):
+        data = np.hstack(inst._data)
```
This will cause a copy. Instead, permute the axes which should yield a view.

choldgraf replied:
done

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1032,71 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+            raise ValueError("Sensor {0} doesn't exist in data".format(key))
+        this_ixs = [i_type[1] for i_type in ch_types if i_type[0] == key][0]
+        this_data = data[this_ixs]
+        scale_factor = np.percentile(np.abs(this_data.ravel()), 99.5)
```
the `np.abs` will also cause a copy. Instead, you could do `np.max(np.percentile(this_data.ravel(), [0.5, 99.5]) * [-1, 1])`

Eric89GXL replied:
(it's not equivalent but it will get you something decent hopefully)

choldgraf replied:
ok...changed to use that version

Eric89GXL commented on mne/viz/tests/test_utils.py:
```diff
@@ -85,4 +86,34 @@ def test_add_background_image():
+    from mne.io import read_raw_fif
+    from mne.event import read_events
+    from mne.epochs import Epochs
+    from ..utils import compute_scalings
```
don't nest these

choldgraf replied:
ok

Eric89GXL commented:
Otherwise LGTM

choldgraf commented:
OK I think I've addressed all these comments - met know if that looks OK

Eric89GXL commented:
Does it still work okay with my suggested change for `np.percentile`?

choldgraf commented:
I'm checking that now

On Tue, May 3, 2016 at 12:55 PM, Eric Larson <notifications@github.com>
wrote:

> Does it still work okay with my suggested change for np.percentile?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3198#issuecomment-216646233>
>

choldgraf commented:
it looks OK on the sample data...I'm trying to think of a situation where that'd break. It's a bit weird to do the * [-1, 1], why not just do `np.max(np.abs(np.percentile(data, [.5, 99.5])))`

Eric89GXL commented:
You can do that instead if you want

choldgraf commented:
ok, updated w/ that convention...that should be an easy-ish fix if this messes up for a particular configuration of data

Eric89GXL commented:
@choldgraf LGTM but now you need to:

1. Update the docstring for our plotting functions that have a `scalings` parameter to say that `'auto'` is an option to use 99.5th percentile.
2. Update an example, if possible, to use it. (You can `git grep "scalings="` from within `doc` and `tutorials` to hopefully find somewhere we've done manually that your auto mode would work nicely wiht).
3. Rebase and update whats_new.py.

@agramfort look good to you otherwise?

dengemann commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1033,72 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+        return scalings
+
+    ch_types = _picks_by_type(inst.info)
+    unique_ch_types = [i_type[0] for i_type in ch_types]
```
this simplifies into:

```unique_ch_types = [ch for ch in ['eeg', 'mag', 'grad', 'ecog'] if ch in inst]```

Thanks the contains mixin!

choldgraf replied:
I'd prefer to leave in the fields for things like stim channels. E.g., I often use this for loading an audio track to visualize next to the ecog channels. The default scaling is 1 (I think) for any non-brain channel types, and I'd like this kind of functionality to deal with this case as well...

dengemann commented on mne/viz/utils.py:
```diff
@@ -1032,3 +1033,72 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
+            continue
+        if key not in unique_ch_types:
+            raise ValueError("Sensor {0} doesn't exist in data".format(key))
+        this_ixs = [i_type[1] for i_type in ch_types if i_type[0] == key][0]
```
can you untuple/unpack this for readability?

something like
```[key_ for key_, value in ch_types if key_ == key][0]```

dengemann replied:
I guess the index at the end is obsolete if you construct properly unique channel types, as suggested?

dengemann replied:
I guess the index at the end is obsolete if you construct properly unique channel types, as suggested?

dengemann commented on mne/viz/tests/test_utils.py:
```diff
@@ -85,4 +89,30 @@ def test_add_background_image():
+    assert_raises(ValueError, _compute_scalings, scalings_def, rand_data)
+    epochs.pick_types(eeg=True, meg=False, copy=False)
+    assert_raises(ValueError, _compute_scalings,
+                  dict([('grad', 'auto')]), epochs)
```
simpler: dict(grad='auto')

choldgraf commented:
hey all - re: docstring, how does this sound?

```
"""
    scalings : dict | 'auto' | None
        Scale factors for the traces. If any fields in scalings are 'auto',
        the scaling factor for that type will be calculated automatically
        from its 99.5th percentile. If scalings == 'auto', scaling factors for
        all channel types will be automatically determined.
        If None, defaults to::

            dict(mag=1e-12, grad=4e-11, eeg=20e-6, eog=150e-6, ecg=5e-4,
                 emg=1e-3, ref_meg=1e-12, misc=1e-3, stim=1, resp=1, chpi=1e-4)
"""
```

Eric89GXL commented:
Something like this would be better "...calculated automatically using a heuristic based on the 99.5th percentile of a chunk of the data."

kingjr commented:
How about
`Scaling factors for the traces. If any fields in `scalings` are 'auto', the scaling factor is set to match the 99.5th percentile of the corresponding data. If `scalings` == 'auto',  all `scalings` fields are set to `auto`.

choldgraf commented:
I'm +1 for @kingjr 's docstring

Eric89GXL commented:
But it doesn't always use all of the data, right? Thus yours and his docstrings are both misleading

kingjr commented:
how about " the scaling factor is set to match the 99.5th percentile of a subsample of the corresponding data."

choldgraf commented:
ah yes that's a good point. If it's an Epochs object then it'll load/use all the epochs, but if it's a Raw object it'll load a chunk of data. How about:

"Scaling factors for the traces. If any fields in scalings are 'auto', the scaling factor is set to match the 99.5th percentile of a subset of the corresponding data. If scalings == 'auto', all scalings fields are set to 'auto'."

for raw, and

"Scaling factors for the traces. If any fields in scalings are 'auto', epochs data will be loaded and the scaling factor is set to match the 99.5th percentile of the corresponding data. If scalings == 'auto', all scalings fields are set to 'auto'."

Eric89GXL commented:
That sounds reasonable.

Since you're improving this, WDYT about using up to 100 MB of data to do the percentile? This is guaranteed to be reasonably fast and low memory, and it's easy to describe ("...from up to 100 MB of the data"). For raw it's getting the number of time points to use as something like this, which would give ~40000 points to use for 306 channels:
```
n_times = max(1, 100000000 // (len(epochs.ch_names) * 8)
```
For epochs it would be something like:
```
n_epochs = max(1, 100000000 // (len(epochs.ch_names) * len(epochs.times) * 8)
```
It will be off by a factor of 2 for complex data but that's okay.

choldgraf commented:
Interesting - that is a cool idea. I am happy to throw that in with this PR...we're agreed 100mb is a good number to go with?

Eric89GXL commented:
It's what we use in our raw readers to keep memory consumption down, so it's consistent at least.

choldgraf commented:
ok I'll take some time to implement that + other changes suggested above, need to get me some coffee first :)

choldgraf commented:
just made edits from above and included the 100mb size max etc, let me know if that looks OK. I'll try to find an example to show this now, and will create a change to whatsnew etc next

choldgraf commented:
ok - also rebased onto master, updated whats new / a doc, and push that. There was an example showing plotting from raw array data, which is a good use case for this PR. That said, I think it'd be great to have an actual "visualization tutorial", maybe a project for the summer? Once these PRs are taken care of (including the CFC one), I'd like to help build some documentation along those lines.

agramfort commented:
there is now tutorials for viz see:

http://martinos.org/mne/dev/tutorials.html

and for Raw

http://martinos.org/mne/dev/auto_tutorials/plot_visualize_raw.html

can you add a note there?

choldgraf commented:
ah that's great - I'll add tomorrow

On Wed, May 4, 2016 at 11:36 PM, Alexandre Gramfort <
notifications@github.com> wrote:

> there is now tutorials for viz see:
>
> http://martinos.org/mne/dev/tutorials.html
>
> and for Raw
>
> http://martinos.org/mne/dev/auto_tutorials/plot_visualize_raw.html
>
> can you add a note there?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3198#issuecomment-217086681>
>

choldgraf commented:
looking at this now - I realized that I didn't see it before because there are two tutorials folders...one in `doc`, and the other in the main `mne_python` folder.  I added in a sentence about it in the raw plotting tutorial. Is that enough? (also rebased onto master)

Eric89GXL commented:
Argh, it does seem like the ones in `doc/tutorials` should be moved to `tutorials`. @agramfort is that right?

choldgraf commented:
Is that something you'd like me to take care of in this PR?

Eric89GXL commented:
No, that should probably be separate

choldgraf commented:
ok cool, will hold off on doing anything else in this PR until somebody
tells me to :)

On Thu, May 5, 2016 at 7:24 PM, Eric Larson <notifications@github.com>
wrote:

> No, that should probably be separate
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3198#issuecomment-217332990>
>

Eric89GXL commented on doc/whats_new.rst:
```diff
@@ -67,6 +67,8 @@ Changelog
 
     - Add option for ``first_samp`` in :func:`mne.make_fixed_length_events` by `Jon Houck`_
 
+    - Add ability to auto-scale channel types for `mne.viz.plot_raw` and `mne.viz.plot_epochs` and corresponding object plotting methods by `Chris Holdgraf`_
```
:func: in front (see other lines above for examples)

agramfort replied:
+1 then +1 for merge

agramfort commented:
>
> Argh, it does seem like the ones in doc/tutorials should be moved to
> tutorials. @agramfort <https://github.com/agramfort> is that right?
>
the files in doc/tutorials are in rst format.

the files in tutorials are python files that are rendered with sphinx
gallery.

it's on purpose.

Eric89GXL commented:
> it's on purpose.

Why not have one or the other? I don't see any reason why these couldn't be made into proper Python `tutorials`:

https://github.com/mne-tools/mne-python/blob/master/doc/tutorials/report.rst
https://github.com/mne-tools/mne-python/blob/master/doc/tutorials/seven_stories_about_mne.rst

The only one that seems like it can't is the MNE command-line tools one, and we can move that elsewhere.

Eric89GXL commented:
At the very least I think we should just move those files to the root `doc` directory, and get rid of the `doc/tutorials` directory -- it's confusing to have that, and one at the repo root.

Eric89GXL commented:
I'll fix the :func: in master -- thanks @choldgraf!

Eric89GXL commented:
Merged by rebase with minor fix

Eric89GXL commented:
After going to edit the doc, I see what you're saying @agramfort, I'm going to leave it alone :)

choldgraf commented:
wahoo! thanks all
